vm/loader/page_table/src/x64.rs

```diff
@@ -299,289 +392,311 @@ impl PteOps for PageTableBuilder {
-    //      1 PDPTE (Level 3)
-    //      4 or 8 PDE tables (Level 2)
-    // Note that there are no level 1 page tables, as 2MB pages are used.
-    let leaf_page_table_count = match identity_map_size {
-        IdentityMapSize::Size4Gb => 4,
-        IdentityMapSize::Size8Gb => 8,
-    };
-    let page_table_count = leaf_page_table_count + if address_bias == 0 { 2 } else { 1 };
-    let mut page_table: Vec<PageTable> = vec![PageTable::new_zeroed(); page_table_count];
-    let mut page_table_allocator = page_table.iter_mut().enumerate();
-
-    // Allocate single PDPTE table.
-    let pdpte_table = if address_bias == 0 {
-        // Allocate single PML4E page table.
-        let (_, pml4e_table) = page_table_allocator
-            .next()
-            .expect("pagetable should always be available, code bug if not");
-
-        // PDPTE table is the next pagetable.
-        let (pdpte_table_index, pdpte_table) = page_table_allocator
-            .next()
-            .expect("pagetable should always be available, code bug if not");
-
-        // Set PML4E entry linking PML4E to PDPTE.
-        let output_address = page_table_gpa + pdpte_table_index as u64 * X64_PAGE_SIZE;
-        pml4e_table[0].set_entry(PageTableEntryType::Pde(output_address));
-
-        // Set PML4E entry to link the additional entry if specified.
-        if let Some((link_target_gpa, linkage_gpa)) = pml4e_link {
-            assert!((linkage_gpa & 0x7FFFFFFFFF) == 0);
-            pml4e_table[linkage_gpa as usize >> 39]
-                .set_entry(PageTableEntryType::Pde(link_target_gpa));
-        }
+}
 
-        pdpte_table
-    } else {
-        // PDPTE table is the first table, if no PML4E.
-        page_table_allocator
-            .next()
-            .expect("pagetable should always be available, code bug if not")
-            .1
-    };
-
-    // Build PDEs that point to 2 MB pages.
-    let top_address = match identity_map_size {
-        IdentityMapSize::Size4Gb => 0x100000000u64,
-        IdentityMapSize::Size8Gb => 0x200000000u64,
-    };
-    let mut current_va = 0;
-
-    while current_va < top_address {
-        // Allocate a new PDE table
-        let (pde_table_index, pde_table) = page_table_allocator
-            .next()
-            .expect("pagetable should always be available, code bug if not");
-
-        // Link PDPTE table to PDE table (L3 to L2)
-        let pdpte_index = get_amd64_pte_index(current_va, 2);
-        let output_address = page_table_gpa + pde_table_index as u64 * X64_PAGE_SIZE;
-        let pdpte_entry = &mut pdpte_table[pdpte_index as usize];
-        assert!(!pdpte_entry.is_present());
-        pdpte_entry.set_entry(PageTableEntryType::Pde(output_address));
-
-        // Set all 2MB entries in this PDE table.
-        for entry in pde_table.iter_mut() {
-            entry.set_entry(PageTableEntryType::Leaf2MbPage(current_va + address_bias));
-            current_va += X64_LARGE_PAGE_SIZE;
+pub struct IdentityMapBuilder<'a> {
```

Comment:
is there a TODO/FUTURE to unify this code with above? Should we instead be having IdentityMapBuilder create a full builder internally and use that?